src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/FilePathNormalizer.cs

```diff
@@ -12,60 +12,187 @@ internal static class FilePathNormalizer
+        var normalizedSpan1 = NormalizeCoreAndGetSpan(filePathSpan1, array1);
+
+        using var _2 = ArrayPool<char>.Shared.GetPooledArray(filePathSpan2.Length, out var array2);
+        var normalizedSpan2 = NormalizeCoreAndGetSpan(filePathSpan2, array2);
 
-        var directory = normalizedPath[..(lastSeparatorIndex + 1)];
-        return directory;
+        return normalizedSpan1.Equals(normalizedSpan2, FilePathComparison.Instance);
     }
 
-    public static bool FilePathsEquivalent(string filePath1, string filePath2)
+    private static ReadOnlySpan<char> NormalizeCoreAndGetSpan(ReadOnlySpan<char> source, Span<char> destination)
     {
-        var normalizedFilePath1 = Normalize(filePath1);
-        var normalizedFilePath2 = Normalize(filePath2);
+        var (start, length) = NormalizeCore(source, destination);
+        return destination.Slice(start, length);
+    }
 
-        return FilePathComparer.Instance.Equals(normalizedFilePath1, normalizedFilePath2);
+    /// <summary>
+    ///  Normalizes the given <paramref name="source"/> file path and writes the result in <paramref name="destination"/>.
+    /// </summary>
+    /// <param name="source">The span to normalize.</param>
+    /// <param name="destination">The span to write to.</param>
+    /// <returns>
+    ///  Returns a tuple containing the start index and length of the normalized path within <paramref name="destination"/>.
+    /// </returns>
+    private static (int start, int length) NormalizeCore(ReadOnlySpan<char> source, Span<char> destination)
+    {
+        if (source.IsEmpty)
+        {
+            if (destination.Length < 1)
+            {
+                throw new ArgumentException("Destination length must be at least 1 if the source is empty.", nameof(destination));
+            }
+
+            destination[0] = '/';
+
+            return (start: 0, length: 1);
+        }
+
+        if (destination.Length < source.Length)
+        {
+            throw new ArgumentException("Destination length must be greater or equal to the source length.", nameof(destination));
+        }
+
+        int charsWritten;
+
+        // Note: We check for '%' characters before calling UrlDecoder.Decode to ensure that we *only*
+        // decode when there are '%XX' entities. So, calling Normalize on a path and then calling Normalize
+        // on the result will not call Decode twice.
+        if (source.Contains("%".AsSpan(), StringComparison.Ordinal))
+        {
+            UrlDecoder.Decode(source, destination, out charsWritten);
+        }
+        else
+        {
+            source.CopyTo(destination);
+            charsWritten = source.Length;
+        }
+
+        // Replace slashes in our normalized span.
+        destination[..charsWritten].Replace('\\', '/');
+
+        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) &&
+            destination is ['/', ..] &&
+            destination is not ['/', '/', ..])
```

Comment:
Maybe too fancy, but I think this can be combined to `destination is ['/', not '/', ..]` ?

Reply:
Oh, that's adorable. :smile:

Reply:
We're slowly but surely converting you to the side of patterns @davidwengier :laughing:

Reply:
I have always been on the side of patterns, but only issue is trying to get my brain to start using them from the get go, rather than seeing how things can be changed to them later. Except positional patterns, I may never get used to them 😛

Reply:
@davidwengier / @333fred : I was mistaken and these patterns matches can't be consolidated as suggested. The problem is that is won't match a span of length 1, which results in an `IndexOutOfRangeException` later.

```C#
var destination = "/".AsSpan();

var firstMatch = destination is ['/', ..] && destination is not ['/', '/', ..];
var secondMatch = destination is ['/', not '/', ..];

Console.WriteLine(firstMatch);
Console.WriteLine(secondMatch);
```

Output:

```
True
False
```

Reply:
The important point to remember is that adding more elements to match or 'not' match will result in a length check: https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQBKApgIbADyAdgDYCeAygA6M2AHgDGAC0YAnAHzoAzgEp8hYqoCWAM3TU56NboDaAcgD0RgDTo2AewAu6UxfQA6ZwF0FK1UQJJv/zAB2AG4vbwBfMMikcKA==.